### ISSUE TITLE
Consider encoding/storing arglists as strings

### ISSUE THREAD

**Issue by lread** [cljdoc/cljdoc#771]
## Currently
API analysis is delivered to cljdoc from cljdoc-analyzer as edn.

## This is fine, but...
`:arglists` are delivered and stored as symbols

Sometimes `:arglists` contain elements that can be written but not read.
A specific example that we have worked around with a custom reader is regular expressions in arglists.

Now that I am exploring #543, I see more examples of potentially writable, but unreadable edn.
For example, the [orchestra lib has an arglists that contains `::some-kw`](https://github.com/jeaye/orchestra/blob/81e5181f7b42e5e2763a2b37db17954f3be0314e/src/cljc/orchestra/detail.cljc#L104-L107). 
This is handled fine for current dynamic analysis, static analysis will not expand to `:some-ns/some-kw` and if we try to `edn/read-string` `::some-kw` it will fail. 

## So...
Why not just store arglists as a list/vector of strings?
This is what clj-kondo analysis does and it has the advantage of being readable with no muss and no fuss. 

Even if we never implement #543, we'd have a better design choice and could drop our custom edn reader.

This means for example, edn `:arglists ([x y] [x y z])`
would become: `:arglists ("[x y]" "[x y z]")`

## The Impact
We sort arglists by arity when rendering the API.
We'll have to ensure this still happens.

We format arglists for rendering with zprint, it expects strings, so we should only need to tweak conversion slightly.

The sqlite database `var` table holds the arglists in a nippy blob in the `meta` column.
We have maybe ~4 million rows in this table (I'll take a peek).
I'll explore how long a migration might take.

The searchset delivered to the client side already converts arglists to strings when converting to json, we'd have to check/update that conversion.

The cljdoc-analyzer will need to deliver the new format of `:arglists`.
I'll work out a migration strategy.

I don't think any other known clients should be affected. 
The Dash offline reader works off rendered HTML, not json.

Please chime in if you, dear reader, can think of any problems with this approach.

## Next Steps
I'll explore, and if things seem to be making sense, I'll follow up with a PR.

**Comment by lread** [comment]
## Update to clj-kondo
I learned that clj-kondo only stored arglists as a vector of strings for original arglists, but not for any overriding metadata. After an [issue](https://github.com/clj-kondo/clj-kondo/issues/2096) and a PR, it now does.

## Suspicious looking arglists handling
### Clj-kondo handling
Clj-kondo does not store suspicious-looking overriding metadata arglists in vector-of-string format. In this case, it keeps the original arglists (if any) as a vector of strings. The winning metadata :arglists, in form format, is still always available under :meta :arglists (which is not necessarily edn-friendly).

### Cljdoc handling
Cljdoc's current behaviour for suspicious-looking arglists is to show them with a warning.
[Example](https://cljdoc.org/d/babbage/babbage/0.1.0/api/babbage.graph#defgraphfn):
![image](https://github.com/cljdoc/cljdoc/assets/967328/dc7c9eb1-c72c-4366-aa08-074f64d6eef6)

One could argue that showing suspicious arglists is not necessary.
A generic warning could suffice and the user could click through to source.
But, the source link will be missing if the jar does not link back to sources in the pom.
So I think it is still probably helpful, we can often get the gist of what the author intended.

### Cljdoc/clj-kondo parity
Cljdoc's current dynamic analysis does not have access to the original arglists, so, even if we wanted to, we cannot easily replicate clj-kondo behaviour of defaulting to the original arglists.

For clj-kondo we can still detect suspicious arglists by looking at the shape of :meta :arglists. In this case, we could convert the form to a string. 

So, if cljdoc sees an :arglist that is a string, it will see this as an indicator that it was deemed suspicious. If it is a vector (of strings), it can display it as valid.

There will be some normal mismatch between dynamic/static processing, and rewrite-clj node-isms  sometimes leak into the clj-kondo :meta :arglists forms. But these should be considered acceptable differences.

**Comment by lread** [comment]
## Db migration exploration

Tested with existing org.clojure/java.jdbc. Too slow. Several minutes for the first 1000 rows.

Tried with `seancorfield/next.jdbc`, much better but still too slow. Current estimate maybe 2.5h to migrate. I'll continue to play around and think about options.

**Comment by lread** [comment]
## Experiment on DigitalOcean Droplet
I'd like some level of comfort that this migration will actually work in production and not take eons. I think running the migration work from a cljdoc docker container in an equivalent DigitalOcean Droplet should suffice.

TL;DR the migration took 2m53s, which is equivalent to my dev box test.

Notes that follow are for future me (or future you).

### Droplet Creation
I created a DigitalOcean droplet that I think matches our production use:
4 GiB	2 vCPUs	4,000 GiB	80 GiB	$0.03571	$24.00
And then I configured it to use Centos7 which matches our current production OS.

### Droplet Setup
I ssh-ed in and ran the following prep cmds:
```Shell
yum update
yum install docker
systemctl start docker
docker pull cljdoc/cljdoc
mkdir -p /data/cljdoc
```
I then used `scp` from my dev box to copy a recent production db backup to `/data/cljdoc`.
then back to my droplet terminal to untar the backup:
```
cd /data/cljdoc
tar xzvf /root/cljdoc-prod-db-2023-06-01.tar.gz
```
I then ran cljdoc within my droplet:
```
cd
docker run -it --entrypoint /bin/bash -v /data/cljdoc:/var/cljdoc:Z cljdoc/cljdoc
```
(For whatever reason, I needed that trailing :Z to get write privs on the mounted volume)

I then edited deps.edn to add next.jdbc, and copied over my migration test (not as a migration but as simple .clj file) and ran it from within the cljdoc docker container.

The duration matched the duration on my dev box.

**Comment by lread** [comment]
## Suspicious looking arglists - existing patterns

To start with, we'll match clj-kondo's strategy of calling any `arglists` that is not a list of vectors as suspicious.

That said, if I look at our production db, I do see arglists that are vectors of vectors, or vectors of lists, or lists of lists. There aren't a ton of these, only ~1300 occurrences across 49 projects. We can loosen what we deem suspicious in the future if any feathers are ruffled.

**Comment by lread** [comment]
## Some Downsides to moving to strings for arg vecs

Some future features like API diffing, for example, would appreciate consistent diffs of arglists. 

Present features, like detecting arglist differences between clj and cljs could also be impacted.

### Commas in maps and comparison

```Clojure
user=> (str {:a 1 :b 2})
"{:a 1, :b 2}"
```
This will happen for dynamic analysis, we'd probably prefer to leave these out.
We can do so by overriding the print-method for IPersistentMap.

For static analysis, whitespace is stripped by clj-kondo.

### Ordering and comparison
When converting to string, map and set order is not deterministic.

For static analysis, the ordering should be as the user typed them.

For dynamic analysis, this will not necessarily be the case, but happens to typically be so in Clojure for small maps and lists.

### When do maps and sets occur in arglists?

Maps and sets can occur as defaults in destructuring, so likely rare.

Maps, though, are the syntax for defaults in destructuring. This somewhat rare, but more common.

Currently, we distinguish clj vs cljs API differences in arglists. 
So if they happen to be ordered differently by our serialization to string, that would be problematic.

### But isn't non-deterministic ordering currently a problem?
Perhaps for consistent rendering.
But small maps and sets typically print in the order they were written.

**Comment by lread** [comment]
Seems to me a decision matrix would help, so here's one:
https://docs.google.com/spreadsheets/d/1A2QouACbwh2x7ub5YhnJpwTWHQ1wdYNsoCVjs7utENk/edit?usp=sharing
Happy to give anyone with an interest in collaborating edit privs, lemme know!

**Comment by lread** [comment]
Ok, the grid helps one to think a bit more. We have the following places we encode arglists:
- in an edn file for API analysis results - requires custom reader
- within a nippy blob in an SQLite database server-side - nippy can encode/decode anything
- to a rendered webpage as HTML
- as json to support searchsets in the client - encodes json-unfriendlies as strings (regexes)
- as json to experimentally explore new UIs

**Comment by lread** [comment]
I stumbled on a good example of [an awkward arglists from clojure itself](https://github.com/clojure/clojure/blob/fc98f92c76254c5a6306debaf0f9df28c3bb3646/src/clj/clojure/core/server.clj#L306-L307).

```Clojure
(defn remote-prepl
  {:added "1.10"}
  [^String host port ^Reader
   in-reader out-fn & {:keys [valf readf] :or {valf read-string, readf #(read %1 false %2)}}]
...)
```


For our current dynamic analysis, this would be evaluated and result in an `:arglists` something like:

```Clojure
([host port in-reader out-fn & {:keys [valf readf]
  :or {valf read-string 
       readf (fn* [p1__5154# p2__5155#]
               (read p1__5154# false p2__5155#))}}])
```
This is awkward for docs, but the author can compensate with an `:arglists` override or... maybe just not do such things.

For static analysis from kondo we'd get an `:arglist-strs` like:
```Clojure
["[host port in-reader out-fn & {:keys [valf readf] :or {valf read-string readf #(read %1 false %2)}}]"]
```
This matches what the author typed, but if we tried to `edn/read-string` it, it would fail.

(side note: interesting that the type hints are not retained... )
